File: clients/csharp/src/Hook0/Runtime.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace Hook0;

/// <summary>What the API answered one request: the status, and the bytes.</summary>
/// <param name="Status">What the API answered under.</param>
/// <param name="Payload">The body it answered, bounded by what the transport agreed to read.</param>
public readonly record struct TransportAnswer(int Status, byte[] Payload);

/// <summary>What the API answered one request, headers included.</summary>
/// <param name="Status">What the API answered under.</param>
/// <param name="Headers">
/// What it answered beside its body, under lowercased names so a caller reads a header without
/// knowing which case the server wrote it in.
/// </param>
/// <param name="Payload">The body it answered.</param>
public readonly record struct TransportDelivery(
    int Status,
    IReadOnlyDictionary<string, string> Headers,
    byte[] Payload);

/// <summary>What a generated method waits on to issue one request.</summary>
public interface ITransport
{
    /// <summary>Issues one request and waits for the answer.</summary>
    /// <param name="method">The HTTP method the operation is issued under.</param>
    /// <param name="path">Where the request lands, absolute or under the base URL.</param>
    /// <param name="query">The name and value pairs of the query string.</param>
    /// <param name="body">What to send as a JSON document, or nothing at all.</param>
    /// <returns>The status and the body the API answered.</returns>
    TransportAnswer Request(
        string method,
        string path,
        IReadOnlyList<KeyValuePair<string, string>> query,
        object? body);
}

/// <summary>What a generated method awaits to issue one request.</summary>
public interface IAsyncTransport
{
    /// <summary>Issues one request and awaits the answer.</summary>
    /// <param name="method">The HTTP method the operation is issued under.</param>
    /// <param name="path">Where the request lands, absolute or under the base URL.</param>
    /// <param name="query">The name and value pairs of the query string.</param>
    /// <param name="body">What to send as a JSON document, or nothing at all.</param>
    /// <param name="cancellationToken">What abandons the request before it is answered.</param>
    /// <returns>The status and the body the API answered.</returns>
    Task<TransportAnswer> RequestAsync(
        string method,
        string path,
        IReadOnlyList<KeyValuePair<string, string>> query,
        object? body,
        CancellationToken cancellationToken);
}

/// <summary>What the API answered is not what it declares it answers.</summary>
public sealed class DecodeException : Exception
{
    /// <summary>Reports a body this client could not read as what the API declares.</summary>
    /// <param name="detail">What could not be read, and out of what.</param>
    public DecodeException(string detail)
        : base(detail)
    {
    }

    /// <summary>Reports a body this client could not read as what the API declares.</summary>
    /// <param name="detail">What could not be read, and out of what.</param>
    /// <param name="cause">What the reader reported.</param>
    public DecodeException(string detail, Exception cause)
        : base(detail, cause)
    {
    }
}

/// <summary>
/// What the generated half of this package reads and writes values through.
/// </summary>
/// <remarks>
/// Everything here is hand-written and never regenerated. It is the one seam between what the API
/// declares — the records, the problems and the methods the generator writes under
/// <c>Generated</c> — and what it does not: how a JSON document becomes a value, how a value
/// becomes a path segment or a query pair, and what happens to a document that does not say what it
/// was declared to say.
/// </remarks>
public static class Runtime
{
    /// <summary>
    /// Longest fragment of a response body a message carries. Bodies are answered by a server this
    /// package does not control, so they are cut at a fixed budget rather than echoed whole into
    /// whatever the caller logs.
    /// </summary>
    public const int MaxPreviewBytes = 256;

    /// <summary>
    /// Largest JSON document read out of a response body, in bytes. The transport caps what it
    /// reads off a socket; this caps what is handed to the reader whichever way the bytes arrived.
    /// </summary>
    public const int MaxPayloadBytes = 8 * 1024 * 1024;

    /// <summary>
    /// Deepest a JSON document may nest before the reader gives up, which is what keeps a document
    /// that is nothing but brackets from growing the stack.
    /// </summary>
    public const int MaxPayloadNesting = 64;

    /// <summary>Most placeholders one path template is filled from.</summary>
    public const int MaxPathParameters = 32;

    /// <summary>Most pairs one query string carries.</summary>
    public const int MaxQueryParameters = 64;

    /// <summary>The characters a path segment carries as themselves.</summary>
    private const string Unreserved =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";

    /// <summary>How a moment travels: RFC 3339, to the precision it carries.</summary>
    private const string MomentFormat = "yyyy-MM-dd'T'HH:mm:ss.FFFFFFFK";

    /// <summary>How a day travels.</summary>
    private const string DayFormat = "yyyy-MM-dd";

    private static readonly JsonSerializerOptions Reading = new()
    {
        MaxDepth = MaxPayloadNesting,
        PropertyNameCaseInsensitive = false,
    };

    // Nothing is dropped wholesale on the way out. Which members may be left out is what the
    // document says and what the generated declarations carry, one member at a time: a blanket
    // "omit what is null" would also omit a member the document requires that happens to carry
    // nothing, and the value would not read back as the value that was written.
    private static readonly JsonSerializerOptions Writing = new()
    {
        MaxDepth = MaxPayloadNesting,
    };

    /// <summary>How a value the API declares is read back out of what it answered.</summary>
    public static JsonSerializerOptions ReadingOptions => Reading;

    /// <summary>How a value is written back the way the API reads it.</summary>
    public static JsonSerializerOptions WritingOptions => Writing;

    /// <summary>Reads a value the API declares out of the body it answered.</summary>
    /// <typeparam name="TValue">What the operation declares it answers.</typeparam>
    /// <param name="payload">The body the transport read.</param>
    /// <returns>The value the body carried.</returns>
    /// <exception cref="DecodeException">
    /// When the body is larger than this package reads, is not JSON, or is not what the API
    /// declares it answers.
    /// </exception>
    public static TValue Read<TValue>(byte[] payload)
    {
        ArgumentNullException.ThrowIfNull(payload);
        if (payload.Length > MaxPayloadBytes)
        {
            throw new DecodeException(
                $"the response is {payload.Length} bytes, above the {MaxPayloadBytes} accepted");
        }

        TValue? read;
        try
        {
            read = JsonSerializer.Deserialize<TValue>(payload, Reading);
        }
        catch (JsonException unreadable)
        {
            throw new DecodeException(
                $"the response is not what the API declares: {Preview(payload)}",
                unreadable);
        }

        return read ?? throw new DecodeException(
            $"the response carries no {typeof(TValue).Name}: {Preview(payload)}");
    }

    /// <summary>Reads a value the API declares, or nothing at all when the body is not one.</summary>
    /// <typeparam name="TValue">What the body is being read as.</typeparam>
    /// <param name="payload">The body the transport read.</param>
    /// <returns>The value the body carried, or <c>null</c> when it carried none.</returns>
    public static TValue? ReadOrNothing<TValue>(byte[] payload)
        where TValue : class
    {
        try
        {
            return Read<TValue>(payload);
        }
        catch (DecodeException)
        {
            return null;
        }
    }

    /// <summary>Writes a value back the way the API reads it.</summary>
    /// <param name="body">What to send.</param>
    /// <returns>The bytes of the JSON document.</returns>
    public static byte[] Write(object body) => JsonSerializer.SerializeToUtf8Bytes(body, Writing);

    /// <summary>As much of a response body as a message may carry.</summary>
    /// <param name="payload">The body the transport read.</param>
    /// <returns>The opening of it, as text.</returns>
    public static string Preview(byte[] payload)
    {
        ArgumentNullException.ThrowIfNull(payload);

        int kept = Math.Min(payload.Length, MaxPreviewBytes);
        string rendered = Encoding.UTF8.GetString(payload, 0, kept);
        return payload.Length > MaxPreviewBytes ? rendered + "…" : rendered;
    }

    /// <summary>What to say about an answer the API document does not describe.</summary>
    /// <param name="status">What the API answered under.</param>
    /// <param name="payload">The body it answered.</param>
    /// <returns>What a caller is told.</returns>
    public static string Unreadable(int status, byte[] payload) =>
        $"the API answered {status.ToString(CultureInfo.InvariantCulture)} with a body this client " +
        $"cannot read: {Preview(payload)}";

    /// <summary>What to say about a problem the API reported.</summary>
    /// <param name="status">What the API answered under.</param>
    /// <param name="problem">The problem document it answered.</param>
    /// <returns>What a caller is told.</returns>
    public static string Reported(int status, object problem) =>
        $"the API answered {status.ToString(CultureInfo.InvariantCulture)}: {problem}";

    /// <summary>Where a request lands, with each placeholder of the template filled in.</summary>
    /// <param name="template">The path as the document writes it, placeholders included.</param>
    /// <param name="filled">The value each placeholder carries.</param>
    /// <returns>The path the request is issued against.</returns>
    public static string Path(string template, IReadOnlyList<(string Name, object? Value)> filled)
    {
        ArgumentNullException.ThrowIfNull(template);
        ArgumentNullException.ThrowIfNull(filled);
        if (filled.Count > MaxPathParameters)
        {
            throw new ArgumentOutOfRangeException(
                nameof(filled),
                filled.Count,
                $"a path is filled from at most {MaxPathParameters} placeholders");
        }

        string written = template;
        foreach ((string name, object? value) in filled)
        {
            written = written.Replace($"{{{name}}}", PathSegment(value), StringComparison.Ordinal);
        }

        return written;
    }

    /// <summary>A value as one segment of a path, with nothing left in it that names another one.</summary>
    /// <param name="value">What the caller passed.</param>
    /// <returns>The segment, percent-encoded.</returns>
    public static string PathSegment(object? value)
    {
        byte[] bytes = Encoding.UTF8.GetBytes(Written(value));
        StringBuilder segment = new(bytes.Length);
        foreach (byte written in bytes)
        {
            char character = (char)written;
            if (Unreserved.Contains(character, StringComparison.Ordinal))
            {
                segment.Append(character);
                continue;
            }

            segment.Append(CultureInfo.InvariantCulture, $"%{written:X2}");
        }

        return segment.ToString();
    }

    /// <summary>
    /// What travels in the query string: everything the document requires, and everything it does
    /// not that the caller actually passed.
    /// </summary>
    /// <param name="required">Name and value pairs the operation always sends.</param>
    /// <param name="optional">Name and value pairs it sends only when they carry something.</param>
    /// <returns>The pairs, in the order they are written.</returns>
    public static IReadOnlyList<KeyValuePair<string, string>> Query(
        IReadOnlyList<(string Name, object? Value)> required,
        IReadOnlyList<(string Name, object? Value)> optional)
    {
        ArgumentNullException.ThrowIfNull(required);
        ArgumentNullException.ThrowIfNull(optional);
        if (required.Count + optional.Count > MaxQueryParameters)
        {
            throw new ArgumentOutOfRangeException(
                nameof(optional),
                required.Count + optional.Count,
                $"a query string carries at most {MaxQueryParameters} pairs");
        }

        List<KeyValuePair<string, string>> pairs = new(required.Count + optional.Count);
        foreach ((string name, object? value) in required)
        {
            pairs.Add(new KeyValuePair<string, string>(name, Written(value)));
        }

        foreach ((string name, object? value) in optional)
        {
            if (value is not null)
            {
                pairs.Add(new KeyValuePair<string, string>(name, Written(value)));
            }
        }

        return pairs;
    }

    /// <summary>How a value travels in a request line, which is not always how C# prints it.</summary>
    /// <param name="value">What the caller passed.</param>
    /// <returns>The text that travels.</returns>
    public static string Written(object? value) => value switch
    {
        null => string.Empty,
        string text => text,
        bool flag => flag ? "true" : "false",
        DateTimeOffset moment => moment.ToUniversalTime()
            .ToString(MomentFormat, CultureInfo.InvariantCulture),
        DateTime moment => moment.ToUniversalTime()
            .ToString(MomentFormat, CultureInfo.InvariantCulture),
        DateOnly day => day.ToString(DayFormat, CultureInfo.InvariantCulture),
        Guid identifier => identifier.ToString("D", CultureInfo.InvariantCulture),
        double number => number.ToString("R", CultureInfo.InvariantCulture),
        float number => number.ToString("R", CultureInfo.InvariantCulture),
        IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
        _ => value.ToString() ?? string.Empty,
    };
}

File: clients/python/src/hook0/runtime.py
"""What the generated half of this package reads and writes values through.

Everything here is hand-written and never regenerated. It is the one seam between what the API
declares — the classes, the problems and the methods the generator writes below this module — and
what it does not: how a JSON document is turned into a typed value, and what happens to a document
that does not say what it was declared to say.

Reading is deliberately strict. A field the document declares as a string and answers as a number
stops the read with the name of the field, rather than yielding an object whose type annotations
lie about what it holds. Every failure of that kind is a `DecodeError`, so a caller has one thing
to catch whatever the shape of the answer was.
"""

from __future__ import annotations

import datetime
import json
import urllib.parse
import uuid
from collections.abc import Callable, Mapping
from enum import StrEnum
from typing import Any, TypeVar

# Longest fragment of a response body an error message carries. Bodies are answered by a server
# this package does not control, so they are cut at a fixed budget rather than echoed whole into
# whatever the caller logs.
MAX_PREVIEW_BYTES = 256

# Largest JSON document read out of a response body, in bytes. The transports cap what they read
# off a socket; this caps what is handed to the parser whichever way the bytes arrived.
MAX_PAYLOAD_BYTES = 8 * 1024 * 1024

Value = TypeVar("Value")

Reader = Callable[[Any], Value]


class DecodeError(ValueError):
    """What the API answered is not what it declares it answers."""


def preview(payload: bytes) -> str:
    """As much of a response body as a message may carry."""
    kept = payload[:MAX_PREVIEW_BYTES]
    rendered = kept.decode("utf-8", errors="replace")
    if len(payload) > MAX_PREVIEW_BYTES:
        return f"{rendered}…"
    return rendered


def decode_payload(payload: bytes) -> Any:
    """The JSON document a response body carries."""
    if len(payload) > MAX_PAYLOAD_BYTES:
        raise DecodeError(f"the response is {len(payload)} bytes, above the {MAX_PAYLOAD_BYTES} accepted")
    try:
        return json.loads(payload)
    except (ValueError, UnicodeDecodeError) as unreadable:
        raise DecodeError(f"the response is not JSON: {preview(payload)}") from unreadable


def as_fields(value: Any, owner: str) -> Mapping[str, Any]:
    """The members of an object the document declares, under the name it declares it with."""
    if not isinstance(value, dict):
        raise DecodeError(f"{owner} is not a JSON object")
    return value


def read(fields: Mapping[str, Any], key: str, reader: Reader[Value]) -> Value:
    """A member the document requires, which is therefore missing when it is absent."""
    if key not in fields:
        raise DecodeError(f"`{key}` is required and was not answered")
    return _named(key, fields[key], reader)


def maybe(fields: Mapping[str, Any], key: str, reader: Reader[Value]) -> Value | None:
    """A member the document does not require, absent as readily as answered as null."""
    if fields.get(key) is None:
        return None
    return _named(key, fields[key], reader)


def _named(key: str, value: Any, reader: Reader[Value]) -> Value:
    """Reads a member, saying which member it was that could not be read."""
    try:
        return reader(value)
    except DecodeError as unreadable:
        raise DecodeError(f"`{key}`: {unreadable}") from unreadable


def as_text(value: Any) -> str:
    """A string, refusing what merely spells like one."""
    if not isinstance(value, str):
        raise DecodeError(f"expected a string, got {type(value).__name__}")
    return value


def as_int(value: Any) -> int:
    """A whole number. `True` is an `int` in Python and is refused here all the same."""
    if isinstance(value, bool) or not isinstance(value, int):
        raise DecodeError(f"expected a whole number, got {type(value).__name__}")
    return value


def as_float(value: Any) -> float:
    """A number, whether the document wrote it with a fractional part or not."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise DecodeError(f"expected a number, got {type(value).__name__}")
    return float(value)


def as_bool(value: Any) -> bool:
    """A boolean, refusing the numbers that stand in for one elsewhere."""
    if not isinstance(value, bool):
        raise DecodeError(f"expected a boolean, got {type(value).__name__}")
    return value


def as_uuid(value: Any) -> uuid.UUID:
    """A UUID, as the document spells one."""
    try:
        return uuid.UUID(as_text(value))
    except ValueError as unreadable:
        raise DecodeError(f"expected a UUID, got `{as_text(value)}`") from unreadable


def as_datetime(value: Any) -> datetime.datetime:
    """A moment, as RFC 3339 spells one."""
    try:
        return datetime.datetime.fromisoformat(as_text(value))
    except ValueError as unreadable:
        raise DecodeError(f"expected a date and a time, got `{as_text(value)}`") from unreadable


def as_date(value: Any) -> datetime.date:
    """A day, as ISO 8601 spells one."""
    try:
        return datetime.date.fromisoformat(as_text(value))
    except ValueError as unreadable:
        raise DecodeError(f"expected a date, got `{as_text(value)}`") from unreadable


def as_json(value: Any) -> Any:
    """A value the document does not describe, which is therefore kept as it arrived."""
    return value


def as_enum(kind: type[StrEnum]) -> Reader[StrEnum]:
    """One of the values a closed list declares."""

    def read_one(value: Any) -> StrEnum:
        try:
            return kind(as_text(value))
        except ValueError as unreadable:
            raise DecodeError(f"`{as_text(value)}` is not one of the values {kind.__name__} declares") from unreadable

    return read_one


def as_list(reader: Reader[Value]) -> Reader[list[Value]]:
    """Every item of an array, each one read the same way."""

    def read_all(value: Any) -> list[Value]:
        if not isinstance(value, list):
            raise DecodeError(f"expected an array, got {type(value).__name__}")
        return [reader(item) for item in value]

    return read_all


def as_map(reader: Reader[Value]) -> Reader[dict[str, Value]]:
    """Every value of an object whose keys the document leaves open."""

    def read_all(value: Any) -> dict[str, Value]:
        if not isinstance(value, dict):
            raise DecodeError(f"expected an object, got {type(value).__name__}")
        return {as_text(key): reader(item) for key, item in value.items()}

    return read_all


def path_segment(value: Any) -> str:
    """A value as one segment of a path, with nothing left in it that could name another one."""
    return urllib.parse.quote(_written(value), safe="")


def query_value(value: Any) -> str:
    """A value as one entry of a query string, escaped when the query is assembled."""
    return _written(value)


def _written(value: Any) -> str:
    """How a value travels in a request line, which is not always how Python prints it."""
    if isinstance(value, bool):
        # `str(True)` is `True`, which no JSON reader on the other end accepts.
        return "true" if value else "false"
    if isinstance(value, StrEnum):
        return value.value
    return str(value)
